### PR TITLE
Organization secret code not copied to clipboard

### LINKED ISSUE #13434: Organization secret code not copied to clipboard
**Describe the bug**

When copying the organization secret code with the special button, it says "Copied to clipboard!" but doesn't work in reality. 

**To Reproduce**

1. Go To Edit Profile 
2. Go to Organization
3. Create an org (or switch to an existing one)
4. Under Grow the team - Your secret code is: --> Copy Button 
image
5. It does not copy the secret code

**Expected behavior**

Clicking on the button just copy the secret code. Or select it if it's not possible to copy. 

**Screenshots**

image
*Not copied in reality

Desktop (please complete the following information):

OS, version: Microsoft, Windows 10
Browser, version: Chrome, latest (88.0.4324.190)

Smartphone (please complete the following information):

Not specific to smartphones.

**Additional context**

I add that we can copy the secret code, just the special button doesn't work.

### PR DESCRIPTION
**Pull request by crhaugen**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This enables copying an organization's secret code.

Having trouble getting the text (or any text) to actually copy to the clipboard.

Not sure how the button should be displayed.  

## Related Tickets & Documents
This pr will fix #13434
## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/javascript/admin/controllers/copy_clipboard_controller.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/* global Runtime */
+import { Controller } from 'stimulus';
+export default class CopyClipboardController extends Controller {
+  handleCopyClick() {
+    const inputValue = document.getElementById('settings-org-secret').value;
+    Runtime.copyToClipboard(inputValue);
```
Comment (crhaugen) [inline app/javascript/admin/controllers/copy_clipboard_controller.js:6]: This doesn't seem to be actually copying anything to the clipboard, I've tried to look at some examples and I know the `Runtime.copyToClipboard(inputValue);` is usually followed by `.then()` but I didn't think I needed to handle anything after doing the copying?

Sorry, I've never really used JS before, so maybe I'm missing something simple.

File: app/views/users/_org_admin.html.erb [hunk 2]
```diff
@@ -60,9 +60,10 @@
     </ol>
     <p>
       Your secret code is:
-      <clipboard-copy for="settings-org-secret" onclick="nextElementSibling.hidden = false">
+      <clipboard-copy for="settings-org-secret" onclick="nextElementSibling.hidden = false" data-controller="copy-clipboard">
         <input type="text" class="ff-monospace ff-s fw-medium" id="settings-org-secret" value="<%= @organization.secret %>" readonly aria-label="Organization secret (to be rotated regularly)">
         <%= inline_svg_tag("copy.svg", aria: true, class: "crayons-icon", title: "Copy to clipboard") %>
+        <button data-action="click->copy-clipboard#handleCopyClick">
```
Comment (crhaugen) [inline app/views/users/_org_admin.html.erb:66]: Not sure how you want to set up how the button looks, this `<button data-action="click->copy-clipboard#handleCopyClick">` just adds another small button to click and the `inline_svg_tag` is where the clipboard image/button is currently is coming from.

File: app/javascript/admin/controllers/copy_clipboard_controller.js (hunk 1 above)
Comment (crhaugen) [reply app/javascript/admin/controllers/copy_clipboard_controller.js:6]: Hi @aitchiss, Thanks for the explanation! I'm still a little unsure how to set up the click even listener in the `app/javascript/packs/clipboardCopy.js`  and call it from the button on `_org_admin.html.erb`

I tried this and having the button call this when clicked but that didn't work
```
handleCopyClick() {
  const inputValue = document.querySelector('[data-copy-text-from-input]');
  document.onclick="nextElementSibling.hidden = false";
  Runtime.copyToClipboard(inputValue);
}
```
and just I tried just using the querySelector and then copying but didn't have any luck with that either. 

`document.querySelector('[data-copy-text-from-input]')`

Comment (aitchiss) [reply app/javascript/admin/controllers/copy_clipboard_controller.js:6]: I think the best way to do this is:

- Over in `app/views/users/_org_admin.html.erb` add an ID to your button (e.g. `copy-button`)
- You will also need to add an ID to the `<input>` element, e.g. 'secret-code-input'
- In the JS, attach the event listener like:

```js
document.getElementById('copy-button').addEventListener('click', handleCopyClick)
```

And then your handleCopyClick function could look something like:

```js
function handleCopyClick() {
  // Get the value from the input you gave an ID to
  const { value } = document.getElementById('secret-code-input');
  // Copy it to the clipboard, then show the span with the success message
  Runtime.copyToClipboard(value).then(() => {
    document.getElementById('copy-text-announcer').classList.remove('hidden');
  })
}
```

I've not tested this myself but it _should_ work (famous last words!). Let me know if you need any more help 👍

**Comment by aitchiss** [comment]
Hi @crhaugen 👋   Just checking in to see how you're getting on with this one - can we offer any more help?

**Comment by citizen428** [comment]
@crhaugen If you don't have time to finish this PR it's ok, just let us know and a team member will take over. Thanks for your contribution!

**Comment by crhaugen** [comment]
Hey sorry, time got away from me, I've tried the latest suggestion for getting the copy to work but still had no luck and it's been difficult to debug. I don't have time anymore for this, so feel free to take it over. 

Thank you for all the help you all have given me! :)

**Comment by aitchiss** [comment]
I'm going to close this PR in favour of https://github.com/forem/forem/pull/16358, which adds a new approach similar to what we discussed. Thanks for all your work on this one @crhaugen  ❤️

**Pull request closed without merging**